---
title: "RNA-seq easy as 1-2-3"
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
output:
  prettydoc::html_pretty:
    fig_caption: no
    fig_height: 25
    fig_width: 30
    number_section: no
    toc: yes
    css: report.css
    theme: architect
---

```{r packages, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(limma)
library(edgeR)
library(ggplot2)
library(Mus.musculus)
library(RColorBrewer)
library(Glimma)
library(gplots)
load(url("http://bioinf.wehi.edu.au/software/MSigDB/mouse_c2_v5p1.rdata"))
```

# Data packaging
## Reading in count data
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
read.delim(files[1], nrow=5)
```

### Sample names
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
x <- readDGE(fileNames, path=inputPath, columns=c(1,3))
samplenames <- substring(colnames(x), 12, nchar(colnames(x)))
samplenames
```

## Organising sample information
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
colnames(x) <- samplenames
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP"))
x$samples$group <- group
lane <- as.factor(rep(c("L004","L006","L008"), c(3,4,2)))
x$samples$lane <- lane
x$samples
```

## Organising gene annotations
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
geneid <- rownames(x)
genes <- select(Mus.musculus, keys=geneid, columns=c("SYMBOL", "TXCHROM"), keytype="ENTREZID")
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
genes <- genes[!duplicated(genes$ENTREZID),]
x$genes <- genes
```
### DGEList
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
x
```

# Data pre-processing
## Transformations from the raw-scale
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
cpm <- cpm(x)
lcpm <- cpm(x, log=TRUE)
L <- mean(x$samples$lib.size) * 1e-6
M <- median(x$samples$lib.size) * 1e-6
c(L, M)
summary(lcpm)
```

## Removing genes that are lowly expressed
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
table(rowSums(x$counts==0)==9)
```

```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
keep.exprs <- filterByExpr(x, group=group)
x <- x[keep.exprs,, keep.lib.sizes=FALSE]
dim(x)
```

```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(x)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
lcpm <- cpm(x, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
```

## Normalising gene expression distributions
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors
```

```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
x2 <- x
x2$samples$norm.factors <- 1
x2$counts[,1] <- ceiling(x2$counts[,1]*0.05)
x2$counts[,2] <- x2$counts[,2]*5
```

```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
par(mfrow=c(1,2))
lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="A. Example: Unnormalised data", ylab="Log-cpm")
x2 <- calcNormFactors(x2)
x2$samples$norm.factors
```

```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
lcpm <- cpm(x2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="B. Example: Normalised data", ylab="Log-cpm")
```

## Unsupervised clustering of samples
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
lcpm <- cpm(x, log=TRUE)
par(mfrow=c(1,2))
col.group <- group
levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
col.group <- as.character(col.group)
col.lane <- lane
levels(col.lane) <-  brewer.pal(nlevels(col.lane), "Set2")
col.lane <- as.character(col.lane)
plotMDS(lcpm, labels=group, col=col.group)
title(main="A. Sample groups")
plotMDS(lcpm, labels=lane, col=col.lane, dim=c(3,4))
title(main="B. Sequencing lanes")
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
glMDSPlot(lcpm, labels=paste(group, lane, sep="_"), groups=x$samples[,c(2,5)],
         launch=FALSE)
```

# Differential expression analysis
## Creating a design matrix and contrasts
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
design <- model.matrix(~0+group+lane)
colnames(design) <- gsub("group", "", colnames(design))
design
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
contr.matrix <- makeContrasts(
   BasalvsLP = Basal-LP,
   BasalvsML = Basal - ML,
   LPvsML = LP - ML,
   levels = colnames(design))
contr.matrix
```

## Removing heteroscedascity from count data
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
v <- voom(x, design, plot=TRUE)
v
```

## Fitting linear models for comparisons of interest
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit)
```

## Examining the number of DE genes
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
summary(decideTests(efit))
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
tfit <- treat(vfit, lfc=1)
dt <- decideTests(tfit)
summary(dt)
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
de.common <- which(dt[,1]!=0 & dt[,2]!=0)
length(de.common)
head(tfit$genes$SYMBOL[de.common], n=20)
```
## Examining individual DE genes from top to bottom
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
basal.vs.lp <- topTreat(tfit, coef=1, n=Inf)
basal.vs.ml <- topTreat(tfit, coef=2, n=Inf)
head(basal.vs.lp)
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
head(basal.vs.ml)
```

## Useful graphical representations of differential expression results
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], xlim=c(-8,13))
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
glMDPlot(tfit, coef=1, status=dt, main=colnames(tfit)[1],
         side.main="ENTREZID", counts=lcpm, groups=group, launch=FALSE)
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
basal.vs.lp.topgenes <- basal.vs.lp$ENTREZID[1:100]
i <- which(v$genes$ENTREZID %in% basal.vs.lp.topgenes)
mycol <- colorpanel(1000,"blue","white","red")
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
heatmap.2(lcpm[i,], scale="row",
   labRow=v$genes$SYMBOL[i], labCol=group,
   col=mycol, trace="none", density.info="none",
   margin=c(8,6), lhei=c(2,10), dendrogram="column")
```

# Gene set testing with camera
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
idx <- ids2indices(Mm.c2,id=rownames(v))
cam.BasalvsLP <- camera(v,idx,design,contrast=contr.matrix[,1])
head(cam.BasalvsLP,5)
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
cam.BasalvsML <- camera(v,idx,design,contrast=contr.matrix[,2])
head(cam.BasalvsML,5)
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
cam.LPvsML <- camera(v,idx,design,contrast=contr.matrix[,3])
head(cam.LPvsML,5)
```
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
barcodeplot(efit$t[,3], index=idx$LIM_MAMMARY_LUMINAL_MATURE_UP,
            index2=idx$LIM_MAMMARY_LUMINAL_MATURE_DN, main="LPvsML")
```
# Software availability
```{r, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
sessionInfo()
```
